src/shaper_srv.erl

```diff
@@ -57,46 +73,35 @@ child_spec(ProcName) ->
      worker,
      [?MODULE]}.
 
--spec start_link(atom()) -> 'ignore' | {'error', _} | {'ok', pid()}.
+-spec start_link(atom()) -> ignore | {error, _} | {ok, pid()}.
 start_link(ProcName) ->
     gen_server:start_link({local, ProcName}, ?MODULE, [], []).
 
--spec worker_prefix() -> string().
-worker_prefix() ->
-    "ejabberd_shaper_".
+-spec worker_name(non_neg_integer()) -> atom().
+worker_name(N) ->
+    ets:lookup_element(ejabberd_shaper_sup, N, 2).
 
-worker_count(_HostType) ->
-    10.
+-spec worker_number(term()) -> non_neg_integer().
+worker_number(Tag) ->
+    erlang:phash2(Tag, worker_count()).
 
--spec worker_names(mongooseim:host_type()) -> [atom()].
-worker_names(HostType) ->
-    [worker_name(HostType, N) || N <- lists:seq(0, worker_count(HostType) - 1)].
-
--spec worker_name(jid:server(), integer()) -> atom().
-worker_name(_Host, N) ->
-    list_to_atom(worker_prefix() ++ integer_to_list(N)).
-
--spec select_worker(mongooseim:host_type(), _) -> atom().
-select_worker(HostType, Tag) ->
-    N = worker_number(HostType, Tag),
-    worker_name(HostType, N).
-
--spec worker_number(mongooseim:host_type(), _) -> non_neg_integer().
-worker_number(HostType, Tag) ->
-    erlang:phash2(Tag, worker_count(HostType)).
+-spec select_worker(term()) -> atom().
+select_worker(Tag) ->
+    N = worker_number(Tag),
+    worker_name(N).
 
 %% @doc Shapes the caller from executing the action.
 -spec wait(HostType :: mongooseim:host_type(),
            Domain :: jid:server(), Action :: atom(),
            FromJID :: jid:jid() | global, Size :: integer()
            ) -> ok | {error, max_delay_reached}.
 wait(HostType, Domain, Action, FromJID, Size) ->
-    gen_server:call(select_worker(HostType, FromJID),
+    gen_server:call(select_worker(FromJID),
                     {wait, HostType, Domain, Action, FromJID, Size}).
 
 %% @doc Ask all shaper servers to forget current shapers and read settings again
-reset_all_shapers(HostType) ->
-    [reset_shapers(ProcName) || ProcName <- worker_names(HostType)].
+reset_all_shapers(_HostType) ->
+    [reset_shapers(ProcName) || {_N, ProcName} <- ets:tab2list(ejabberd_shaper_sup)].
```

Comment:
ideally, operations for storing and getting the list of workers should be in the same file.
i.e. ejabberd_shaper_sup should have a function get_worker_names.